### PR TITLE
document one shot AEAD length restrictions

### LINKED ISSUE #4303: data size limits are not documented for encrypt method on AESGCM
Using the AESGCM primitive, a OverflowError can occur.

Steps to reproduce:

```
import os
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
buf = b'x' * (2**32 + 1)
lend = default_backend()
iv = os.urandom(16)
key = os.urandom(32)
encryptor = AESGCM(key)
byts = encryptor.encrypt(iv, buf, None)
```
Results in the following exception:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/user/.pyenv/versions/somepyenv/lib/python3.6/site-packages/cryptography/hazmat/primitives/ciphers/aead.py", line 147, in encrypt
    backend, self, nonce, data, associated_data, 16
  File "/home/user/.pyenv/versions/somepyenv/lib/python3.6/site-packages/cryptography/hazmat/backends/openssl/aead.py", line 109, in _encrypt
    processed_data = _process_data(backend, ctx, data)
  File "/home/user/.pyenv/versions/somepyenv/lib/python3.6/site-packages/cryptography/hazmat/backends/openssl/aead.py", line 92, in _process_data
    res = backend._lib.EVP_CipherUpdate(ctx, buf, outlen, data, len(data))
OverflowError: integer 4294967297 does not fit '32-bit int'
```

Test environment:
Using Python 3.6.1, cryptography 2.2.2, cffi 1.11.5. Cryptography was installed using pip install onto a venv managed by pyenv.

Suggested course of action:
Update documentation to reflect that the amount of data that can be pushed through `encrypt()` is limited.  This should probably be documented wherever user data could be encrypted and potentially cause a failure for data to cross the cffi interface.

### PR DESCRIPTION
**Pull request by reaperhulk**
These limitations are not enforced in the code outside of cffi raising an overflow if you pass > 32-bit int. Do we want to add tests for that?

fixes #4303

### CODE REVIEW

**Comment by alex** [comment]
If it's practicable (64-bit only I guess?), yes I'd like to add tests for this.

Should we also document that for `AESGCM` encrypting large volumes of data under the same key has problems?

**Comment by alex** [comment]
Why 8GB? Surely it's 4GB+1?

**Comment by reaperhulk** [comment]
All the memory for the data to be encrypted + all the memory for the return buffer.

**Comment by alex** [comment]
Oh, right. 8GB isn't thaaaaat much these days

**Comment by reaperhulk** [comment]
**shakes fist at Alex**

File: tests/hazmat/primitives/test_aead.py [hunk 1]
```diff
@@ -22,6 +22,11 @@
 )
 
 
+class FakeData(object):
+    def __len__(self):
+        return 2 ** 32 + 1
```
[the remark this answers is not in the document]
Comment (reaperhulk) [reply tests/hazmat/primitives/test_aead.py]: Indeed it doesn't. Suggestion?

[the remark this answers is not in the document]
Comment (reaperhulk) [reply tests/hazmat/primitives/test_aead.py]: Unless you have a better suggestion I'm tempted to just revert the test and document the limitation only.

**Comment by reaperhulk** [comment]
I guess it isn’t going to change, so okay.

File: tests/hazmat/primitives/test_aead.py [hunk 2]
```diff
@@ -46,6 +51,17 @@ def test_chacha20poly1305_unsupported_on_older_openssl(backend):
 )
 @pytest.mark.requires_backend_interface(interface=CipherBackend)
 class TestChaCha20Poly1305(object):
+    def test_data_too_large(self, backend):
```
Comment (alex) [inline tests/hazmat/primitives/test_aead.py]: `backend` arg is not used

File: tests/hazmat/primitives/test_aead.py [hunk 3]
```diff
@@ -168,6 +184,17 @@ def test_aesccm_unsupported_on_older_openssl(backend):
 )
 @pytest.mark.requires_backend_interface(interface=CipherBackend)
 class TestAESCCM(object):
+    def test_data_too_large(self, backend):
```
Comment (alex) [inline tests/hazmat/primitives/test_aead.py]: same

File: tests/hazmat/primitives/test_aead.py [hunk 4]
```diff
@@ -309,6 +336,17 @@ def _load_gcm_vectors():
 
 @pytest.mark.requires_backend_interface(interface=CipherBackend)
 class TestAESGCM(object):
+    def test_data_too_large(self, backend):
```
Comment (alex) [inline tests/hazmat/primitives/test_aead.py]: same

**Comment by reaperhulk** [comment]
For some reason I thought those were required to skip, but since we're skipping at the class level they can be removed. Done.

File: docs/hazmat/primitives/aead.rst [hunk 5]
```diff
@@ -55,11 +55,14 @@ also support providing integrity for associated data which is not encrypted.
...
         :param bytes nonce: A 12 byte value. **NEVER REUSE A NONCE** with a
             key.
-        :param bytes data: The data to encrypt.
+        :param bytes data: The data to encrypt. Must be less than or equal to
+            2\ :sup:`32` bytes.
         :param bytes associated_data: Additional data that should be
             authenticated with the key, but does not need to be encrypted. Can
             be ``None``.
         :returns bytes: The ciphertext bytes with the 16 byte tag appended.
+        :raises OverflowError: If ``data`` or ``associated_data`` is greater
```
Comment (alex) [inline docs/hazmat/primitives/aead.rst]: `s/greater/longer/`

File: docs/hazmat/primitives/aead.rst [hunk 6]
```diff
@@ -55,11 +55,14 @@ also support providing integrity for associated data which is not encrypted.
 
         :param bytes nonce: A 12 byte value. **NEVER REUSE A NONCE** with a
             key.
-        :param bytes data: The data to encrypt.
+        :param bytes data: The data to encrypt. Must be less than or equal to
```
Comment (alex) [inline docs/hazmat/primitives/aead.rst]: I'd drop the second sentence.

**Comment by vEpiphyte** [comment]
@reaperhulk @alex thanks!

**Pull request merged**